modules/sharedaddy/sharing-sources.php

```diff
@@ -1602,6 +1602,49 @@ function jetpack_sharing_pocket_init() {
 
 }
 
+class Share_Telegram extends Sharing_Source {
+	public $shortname = 'telegram';
+
+	public function __construct( $id, array $settings ) {
+		parent::__construct( $id, $settings );
+	}
+
+	public function get_name() {
+		return __( 'Telegram', 'jetpack' );
+	}
+	public function process_request( $post, array $post_data ) {
+		// Record stats
+		parent::process_request( $post, $post_data );
+		$telegram_url = esc_url_raw( 'https://telegram.me/share/url?url=' . rawurlencode( $this->get_share_url( $post->ID ) ) . '&text=' . rawurlencode( $this->get_share_title( $post->ID ) ) );
+		wp_redirect( $telegram_url );
+		exit;
+	}
+
+	public function get_display( $post ) {
+		return $this->get_link( $this->get_process_request_url( $post->ID ), _x( 'Telegram', 'share to', 'jetpack' ), __( 'Click to share on Telegram', 'jetpack' ), 'share=telegram' );
+	}
+
+	function display_footer() {
+		$this->js_dialog( $this->shortname, array( 'width' => 450, 'height' => 450 ) );
+	}
+}
+
+class Jetpack_Share_WhatsApp extends Sharing_Source {
+	public $shortname = 'jetpack-whatsapp';
+
+	public function __construct( $id, array $settings ) {
+		parent::__construct( $id, $settings );
+	}
+
+	public function get_name() {
+		return __( 'WhatsApp', 'jetpack' );
+	}
+
+	public function get_display( $post ) {
+		return $this->get_link( 'whatsapp://send?text=' . rawurlencode( $this->get_share_title( $post->ID ) ) . ' ' . rawurlencode( get_permalink( $post->ID ) ), _x( 'WhatsApp', 'share to', 'jetpack' ), __( 'Click to share on WhatsApp', 'jetpack' ) );
```

Comment:
We left the implementation intentional like (the share button visible everywhere) because it also causes a lot of confusion when the share button is enable but don't see it on the desktop side 
because it is only shown on smaller screens. The problem with the responsive approach is at what side do you determine to hide the button? Since the button should be available on any devide that has the app ( now there is even a desktop app https://blog.whatsapp.com/10000621/Introducing-WhatsApps-desktop-app ).

So I am more inclined to let users style the button them self. We should provide some documentation on how to do it but leave the default functionality to be the same as other buttons.